src/sentry/users/web/accounts.py

```diff
@@ -374,3 +390,84 @@ def confirm_email(request: HttpRequest, user_id: int, hash: str) -> HttpResponse
     return HttpResponseRedirect(reverse("sentry-account-settings-emails"))
+
+
+@set_referrer_policy("strict-origin-when-cross-origin")
+@login_required
+@control_silo_function
+def confirm_signed_email(
+    request: HttpRequest, signed_data: str
+) -> HttpResponseRedirect | HttpResponse:
+    EMAIL_CONFIRMATION_SALT = options.get("user-settings.signed-url-confirmation-emails-salt")
+
+    use_signed_urls = options.get("user-settings.signed-url-confirmation-emails")
+    if not use_signed_urls:
+        msg = ERR_CONFIRMING_EMAIL
+        level = messages.ERROR
+        messages.add_message(request, level, msg)
+        return HttpResponseRedirect(reverse("sentry-account-settings-emails"))
+
+    msg = _("Thanks for confirming your email")
+    level = messages.SUCCESS
+
+    try:
+        data = unsign(
+            signed_data, salt=EMAIL_CONFIRMATION_SALT, max_age=172800
+        )  # max age is 2 days in seconds
+
+        # is the currently logged in user the one that
+        # wants to add the email to their account
+        if request.user.id != int(data["user_id"]):
+            raise InvalidRequest
+
+        # check to see if the email has already been verified
+        try:
+            email = UserEmail.objects.get(user=request.user.id, email=data["email"])
+            if email.is_verified:
+                raise VerifiedEmailAlreadyExists()
+        except UserEmail.DoesNotExist:
+            # user email does not exist, so we can create it
+            pass
+    except VerifiedEmailAlreadyExists:
+        msg = WARN_EMAIL_ALREADY_VERIFIED
+        level = messages.INFO
+        messages.add_message(request, level, msg)
+        return HttpResponseRedirect(reverse("sentry-account-settings-emails"))
+    except SignatureExpired:
+        msg = ERR_SIGNATURE_EXPIRED
+        level = messages.ERROR
+        messages.add_message(request, level, msg)
+        return HttpResponseRedirect(reverse("sentry-account-settings-emails"))
+    except (InvalidRequest, BadSignature):
+        msg = ERR_CONFIRMING_EMAIL
+        level = messages.ERROR
+        messages.add_message(request, level, msg)
+        return HttpResponseRedirect(reverse("sentry-account-settings-emails"))
+    except Exception:
+        logger.exception("user.email.signed-confirm.error")
+        msg = ERR_CONFIRMING_EMAIL
+        level = messages.ERROR
+        messages.add_message(request, level, msg)
+        return HttpResponseRedirect(reverse("sentry-account-settings-emails"))
+
+    user = User.objects.get(id=request.user.id)
+    email = UserEmail.objects.create(
+        user=user,
+        email=data["email"],
+        validation_hash="",
+        is_verified=True,
+    )
+    email.save()
+
+    email_verified.send(email=email.email, sender=email)
+    logger.info(
+        "user.email.signed-confirm",
+        extra={
+            "user_id": request.user.id,
+            "ip_address": request.META["REMOTE_ADDR"],
+            "email": email.email,
```

Comment:
Log the `useremail.id` instead of the actual email.
Or maybe a hash of the email?